Data/Graph.hs

```diff
@@ -298,8 +356,35 @@ graphFromEdges' x = (a,b) where
 
 -- | Build a graph from a list of nodes uniquely identified by keys,
 -- with a list of keys of nodes this node should have edges to.
--- The out-list may contain keys that don't correspond to
--- nodes of the graph; they are ignored.
+--
+-- - @node@ is the type of data held at each vertex
+-- - @key@ is uniquely identifies each node in the graph
+--
+-- @(graph, nodeFromVertex, vertexFromKey) = graphFromEdges edgeList@
+--
+-- * @graph@ is the raw, array based adjacency list for the graph.
+-- * @nodeFromVertex@ returns the node associated with the given 0-based @Int@
+--   vertex, see /warning/ below.
+-- * @vertexFromKey@ returns the @Int@ vertex for the key if it exists in the
+--   graph, @Nothing@ otherwise.
+--
+-- Note: The out-list may contain keys that don't correspond to nodes of the
+-- graph; they are ignored.
+--
+-- Warning: The @nodeFromVertex@ function is partial and will cause a runtime
+-- exception if the given @Vertex@ does not exist.
+--
+-- ==== __Examples__
+--
+-- An empty graph.
+-- > (graph, nodeFromVertex, vertexFromKey) = G.graphFromEdges []
+-- > graph = array (0,-1) []
+--
+-- > (graph, nodeFromVertex, vertexFromKey) = G.graphFromEdges [("a", 'a', ['b']), ("b", 'b', ['c']), ("c", 'c', [])]
+-- > graph == array (0,2) [(0,[1]),(1,[2]),(2,[])]
+-- > nodeFromVertex 0 == ("a",'a',"b")
+-- > vertexFromKey 'a' == Just 0
+--
```

Comment:
I don't think you've really explained this very well yet, unfortunately. That's largely because the function is horrible, but we should do our best. It sounds like we're taking an adjacency list representing a graph with vertices of type `key`, whose vertices are labeled with values of type `node`, and producing a `Graph`-based representation of that list. The `Graph` represents the shape, and the functions explains how the original adjacency list fits in. The first function could really just be `Vertex -> (key, node)`, but for convenience it includes the edges from the key. I think. Terrible API.

Reply:
Yeah, the API here is a mess. I would have expected `Graph` to be a type (without its internals exposed) which kept track of the association functions, and then each of the functions in the API would just operate on `Graph`. I would enjoy taking a stab at re-writing this module all together, obviously it would break compatibility so that would have to be resolved, but I'm convinced we can make this much better.

I've updated the comment and provided another example for how to "safely" use the API. Let me know what you think.

Reply:
There's also the question of if the `containers` package should even have a thorough graph library or if that should be left to another package. I'm not sure if there's a "canonical" graph package, although there's a bunch of existing packages which provide graphs with varying levels of functionality/documentation/testing.

Reply:
I think you might have fun modernizing `fgl`. It's a pretty design, but the API is very old-fashioned. I don't know much about graphs myself, so I'm probably not the best one to give detailed advice.

Reply:
I'll reach out to the author :)